### PR TITLE
Implement Cel2 migration tool

### LINKED ISSUE #36: Deploy OP contracts during migration
When migrating the state we need to deploy required smart contracts. For example, some contracts an proxies are expected at certain addresses, which can only done during the migration.

```[tasklist]
### Tasks
- [x] Create a clean optimism "genesis" fork, that includes all necessary smart contracts.
- [x] Extend the migration tool to merge all accounts from that genesis file into the migrated state db.
```

### PR DESCRIPTION
**Pull request by palango**
Part of https://github.com/celo-org/optimism/issues/108
Resolves https://github.com/celo-org/optimism/issues/36

This PR lays the foundation for the Cel2 migration tool.

It works on an existing Celo L1 state database and deploys the OP L2 contracts based on given configuration and L1 contract deployment. The resulting state changes are added in a new block on top of the existing chain, which can then be loaded by `op-geth` in a rollup setting.

### CODE REVIEW

File: op-chain-ops/cmd/op-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/op-migrate/main.go:1]: I'd leave this untouched and just add a celo-migrate. It'll help prevent conflicts

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go:36]: What is the deploy config, where can I get it from?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: What are the deploy config and l1-deployments, where can I get them from?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: There is no `--l1-starting-block` flag

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/main.go]: Also there is no `--deployment-dir` flag.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: When are we choosing these values?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go:186]: I'd put this up the top, so the user doesn't have to wait to find out they made a mistake in calling the command

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: Is this still todo?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: I'd put this up the top, so the user doesn't have to wait to find out they made a mistake in calling the command. Also for consistency I think it's nicer to just handle all the flags first.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go:177]: There doesn't seem to be any reason to do this after retrieving l1StartBlock, since the only relevant check it makes is that L1StartingBlockTag is not nil. I think this would be neater (and faster fail for the user) if the config check came up front, and a function could be extracted to retrieve the l1 block.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: Probably safer to defer this, or you could move the open and close into `ApplyMigrationChangesToDB`

File: op-chain-ops/cmd/op-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/op-migrate/main.go:1]: This doesn't exist in upstream any more. It was added in https://github.com/celo-org/optimism/commit/063933de73ac1c5153ff96bec2190366cabd993a

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: This is what's created before and during the L1 contracts deployment: https://github.com/celo-org/optimism/pull/114/files#diff-da3d22b46956839805580fbc5ebc394e410cd34b9835cf9a7ab8fe4bde5b6447R29-R31

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: I added this task as a follow-up in https://github.com/celo-org/optimism/issues/108

In general, it's one of the parameters we need to decide on. Is there a tracking issue for those already?
Edit: created one in https://github.com/celo-org/celo-blockchain-planning/issues/251

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: I see your point, but until the next rebase I'd like to not change upstream code like this. But I added it as a task to https://github.com/celo-org/optimism/issues/108

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: Moved it inside the function.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: I'm not following, this isn't upstream code?

File: op-chain-ops/cmd/op-migrate/main.go
Comment (piersy) [reply op-chain-ops/cmd/op-migrate/main.go:1]: Oh right 👍  The name (op-migrate) confused me

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: It is upstream code.

And what I wanted to convey is that I'd like to keep it as close to the upstream as possible before the next rebase in order to reduce merge conflicts.

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: I see that `config.Check` is code that exists in upstream, however I'm still somewhat confused about how a conflict would arise here, since this file `op-chain-ops/cmd/celo-migrate/main.go` does not exist in upstream?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]:
```suggestion
```

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]:
```suggestion
	log.Info("Read header from database", "header", header)
```

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: There doesn't seem to be a reason for this `dbFactory()`

```suggestion
	// Set up the backing store.
	underlyingDB := state.NewDatabaseWithConfig(ldb, &trie.Config{
		Preimages: true,
	})

	// Open up the state database.
	db, err := state.New(header.Root, underlyingDB, nil)
	if err != nil {
		return nil, fmt.Errorf("cannot open StateDB: %w", err)
	}
```

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]:
```suggestion
	// Now we iterate through all accounts that have been written there and set them inside the statedb.
```

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go:279]: Do we have a ticket for this, or maybe we don't have to worry if we don't hit any existing accounts? But leaving it to the last minute to find that out is probably not a good idea.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: It would be good to know why the balanceDiff needs to be 0. I notice that the script is not subtracting the balance of overwritten accounts from the balanceDiff, it seems like it should.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: Should this be the number of the transition block?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]:
```suggestion
		TxHash:      types.EmptyTxsHash,
		ReceiptHash: types.EmptyReceiptsHash,
```

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: I believe it means that there is no chance for the block to be re-orged, which in eth2 means it's got 64 slots or whereabouts that build off it.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: Do we need to add cancun here?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: I'm not concerned about the code in the `main.go` file, but the code in the implementation of `Check`.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go:279]: Added a task to https://github.com/celo-org/optimism/issues/108

But my general idea here is that there should not be any accounts that are overwritten. The CREATE2 contract is one example that is, but there the contract code stays the same, so it's fine. But other cases we should find and fix beforehand.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: Any idea what that means for us then? Is there the same concept of finalization with `op-node`/`op-geth`?

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: Yes, good find.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: Added it.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: The property here is that we must not change the amount of celo available.

That said, you have a point with using the aggregate counter. Instead we can just use a per-account check, which makes thing more obvious.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: I guess the reason for this was to make `underlyingDB` private and unusable by accident.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/main.go:177]: Ahh ok, I wasn't suggesting changing Check, just placing it differently in this file.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/main.go]: For us starting out with a centralised sequencer assuming we wait for deposit transactions on the L1 to become finalized (i.e. 64 blocks deep) then the blocks put out by the sequencer can be considered immediately final (assuming you can execute the block and see it is correct). However I'm not sure how the fraud proving factors into that. BTW this is the [PR](https://github.com/ethereum/go-ethereum/pull/24282) where that method was added.

File: op-chain-ops/cmd/celo-migrate/main.go
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/main.go]: It seems like we would want to fail here, if we need to enforce that we don't change the amount of available celo

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/main.go]: I think we don't want to fail during testing now, but we should rather notice those problems (currently this isn't triggered)

**Review by piersy — APPROVED** [review]
Approved, pending task to make test fail on mismatched account balances.

**Pull request merged**
